File: backend/routes/aiRoutes.js
// routes/aiRoutes.js
const {
  updateDbContextFile,
  updateGithubContextFile,
  updateResumeContextFile,
  buildMemoryIndex,
  askLLM,
  askWithRAG,
  // Add references to new controller functions:
  suggestFollowUpQuestions,
  snapshotMemoryUpdate,
  optimizeQuery,
} = require("../controllers/aiContextManager");

async function aiRoutes(fastify, options) {
  fastify.get("/", async (request, reply) => {
    return reply.send({ message: "AI Routes are working!" });
  });

  // Endpoint to manually trigger (re)creation of index and context updates
  fastify.route({
    method: ["GET", "POST"],
    url: "/create-index",
    handler: async (request, reply) => {
      try {
        // Regenerate all context files and the memory index
        await updateDbContextFile();
        await updateGithubContextFile();
        await updateResumeContextFile();
        // Force rebuild memory index to incorporate updated context
        await buildMemoryIndex(true);
        reply.send({
          message: "Context files updated and memory index built successfully.",
        });
      } catch (error) {
        console.error("Error creating index:", error);
        reply.code(500).send({ error: error.message });
      }
    },
  });

  // Endpoint to ask a question to the AI using the indexed context
  fastify.post("/ask-chat", async (request, reply) => {
    try {
      const { query } = request.body;
      if (!query || query.trim().length === 0) {
        return reply.code(400).send({ message: "Query cannot be empty." });
      }
      const answer = await askLLM(query);
      reply.send({ answer });
    } catch (error) {
      console.error("Error handling /ask-chat:", error);
      reply.code(500).send({ error: error.message });
    }
  });

  // **New** endpoint to get suggested follow-up questions
  fastify.post("/suggestFollowUpQuestions", async (request, reply) => {
    try {
      const { query, response, conversationMemory } = request.body;
      if (!query || !response) {
        return reply
          .code(400)
          .send({ error: "Both query and response are required." });
      }
      const suggestions = await suggestFollowUpQuestions(
        query,
        response,
        conversationMemory
      );
      reply.send({ suggestions });
    } catch (error) {
      console.error("Error handling /suggestFollowUpQuestions:", error);
      reply.code(500).send({ error: error.message });
    }
  });

  // **New** endpoint to update conversation memory snapshot
  fastify.post("/snapshotMemoryUpdate", async (request, reply) => {
    try {
      const { previousMemory, query, response } = request.body;
      if (!query || !response) {
        return reply.code(400).send({
          error: "Query and response are required for memory update.",
        });
      }
      const updatedMemory = await snapshotMemoryUpdate(
        previousMemory || "",
        query,
        response
      );
      reply.send({ memory: updatedMemory });
    } catch (error) {
      console.error("Error handling /snapshotMemoryUpdate:", error);
      reply.code(500).send({ error: error.message });
    }
  });

  // New endpoint:
  fastify.post("/optimize-query", async (request, reply) => {
    try {
      const { query, conversationMemory } = request.body;
      if (!query) {
        return reply.code(400).send({ error: "Query is required." });
      }
      const optimizedQuery = await optimizeQuery(
        conversationMemory || "",
        query
      );
      reply.send({ optimizedQuery });
    } catch (err) {
      console.error("Error /optimize-query:", err);
      reply.code(500).send({ error: err.message });
    }
  });
}

module.exports = aiRoutes;

File: Miscellaneous/backend GoLang/routes/airoutes.go
package routes

import (
	"net/http"

	"example.com/portfolio-backend/controllers"
	"github.com/gin-gonic/gin"
)

// RegisterAiRoutes sets up all AI-related endpoints under /api/ai.
func RegisterAiRoutes(router *gin.RouterGroup) {
	// Basic test endpoint
	router.GET("/", func(c *gin.Context) {
		c.JSON(200, gin.H{"message": "AI Routes are working!"})
	})
	// Trigger manual (re)creation of context index and memory index
	router.GET("/create-index", handleCreateIndex)
	router.POST("/create-index", handleCreateIndex)
	// Ask a question to the AI using indexed context
	router.POST("/ask-chat", func(c *gin.Context) {
		var req struct {
			Query string `json:"query"`
		}
		if err := c.BindJSON(&req); err != nil || req.Query == "" {
			c.JSON(http.StatusBadRequest, gin.H{"message": "Query cannot be empty"})
			return
		}
		answer, err := controllers.AskLLM(req.Query)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		} else {
			c.JSON(http.StatusOK, gin.H{"answer": answer})
		}
	})
	// Get suggested follow-up questions
	router.POST("/suggestFollowUpQuestions", func(c *gin.Context) {
		var req struct {
			Query              string `json:"query"`
			Response           string `json:"response"`
			ConversationMemory string `json:"conversationMemory"`
		}
		if err := c.BindJSON(&req); err != nil || req.Query == "" || req.Response == "" {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Both query and response are required"})
			return
		}
		suggestions, err := controllers.SuggestFollowUpQuestions(req.Query, req.Response, req.ConversationMemory)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		} else {
			c.JSON(http.StatusOK, gin.H{"suggestions": suggestions})
		}
	})
	// Update conversation memory snapshot
	router.POST("/snapshotMemoryUpdate", func(c *gin.Context) {
		var req struct {
			PreviousMemory string `json:"previousMemory"`
			Query          string `json:"query"`
			Response       string `json:"response"`
		}
		if err := c.BindJSON(&req); err != nil || req.Query == "" || req.Response == "" {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Query and response are required"})
			return
		}
		updatedMemory, err := controllers.SnapshotMemoryUpdate(req.PreviousMemory, req.Query, req.Response)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		} else {
			c.JSON(http.StatusOK, gin.H{"memory": updatedMemory})
		}
	})
	// Optimize a query for better retrieval
	router.POST("/optimize-query", func(c *gin.Context) {
		var req struct {
			Query              string `json:"query"`
			ConversationMemory string `json:"conversationMemory"`
		}
		if err := c.BindJSON(&req); err != nil || req.Query == "" {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Query is required"})
			return
		}
		optimized, err := controllers.OptimizeQuery(req.ConversationMemory, req.Query)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		} else {
			c.JSON(http.StatusOK, gin.H{"optimizedQuery": optimized})
		}
	})
}

// handleCreateIndex triggers context file updates and memory index rebuild.
func handleCreateIndex(c *gin.Context) {
	if err := controllers.InitContext(); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
	} else {
		c.JSON(http.StatusOK, gin.H{"message": "Context files updated and memory index built successfully."})
	}
}
